### DISCUSSION TITLE
Is it possible to use Decompose with SwiftUI NavigationLink?

### DISCUSSION THREAD

**Discussion by plusmobileapps** [badoo/Decompose#203] [Q&A] [answered]
I was integrating Decompose into an iOS app and was wondering if it's possible to use the [`NavigationLink`](https://developer.apple.com/documentation/swiftui/navigationlink) from SwiftUI to achieve the native back swipe gesture and toolbar? I applied a simple example to the todo app using `NavigationLink`.  

![ios-navigationlink mov](https://user-images.githubusercontent.com/31364841/128314969-37e50b76-90bd-4919-907d-a78b8aec6c69.gif)

```swift
struct MainView: View {
    private let component: TodoMain
    
    @ObservedObject
    private var models: ObservableValue<TodoMainModel>
    
    init(_ component: TodoMain) {
        self.component = component
        self.models = ObservableValue(component.models)
    }
    
    var body: some View {
        let model = models.value
        
        return NavigationView {
            VStack {
                // how would the new component get to the destination parameter? 
                NavigationLink(destination: Text("View2").navigationBarTitle("Title View2", displayMode: .inline)) {
                      Text("NavigationLink") 
                  }
                ListView(
                    items: model.items,
                    onItemClicked: component.onItemClicked,
                    onDoneChanged: component.onItemDoneChanged
                )

                InputView(
                    textBinding: Binding(get: { model.text }, set: component.onInputTextChanged),
                    onAddClicked: component.onAddItemClicked
                )
            }.navigationBarTitle("Todo Sample", displayMode: .inline)
        }
    }
}
```

Using the code above, I understand that the current [`RootView`](https://github.com/JetBrains/compose-jb/blob/master/examples/todoapp/ios/ios/RootView.swift#L20) supplies the components to each individual component and was not quite how to supply the new component to the `NavigationLink`? 🤔  

Any help would be greatly appreciated or if there is a different way of achieving the native gesture without `NavigationLink`

**Comment by arkivanov** [comment, marked as answer]
First of all, I'm not quite familiar with SwiftUI. As far as I know, `NavgiationLink` should be put around each item. I found the following links, which look useful:

- https://www.hackingwithswift.com/quick-start/swiftui/how-to-push-a-new-view-when-a-list-row-is-tapped
- https://www.simpleswiftguide.com/how-to-add-navigationview-to-list-in-swiftui-and-show-detail-view-using-navigationlink/

But as far as I know, `NavigationLink` is active component. It accepts a `View`, and SwiftUI displays it inside `NavigationView`. Looks like in this case the navigation is managed by SwiftUI, however ideally it should be managed by Decompose. It's like still using Fragments in Android. So if possible, I would avoid `NavigationLink` and use `NavigationView` instead, as in the TodoApp example.

Also when I was playing with `NavigationLink` in MVIKotlin, I found that it creates views eagerly. So for N items in the list (and so N `NavigationLink`s), it was immediately creating N detail views. So I found a workaround with [LazyView](https://github.com/arkivanov/MVIKotlin/blob/bad123550ff7e451708280efbc4394ffdc347107/sample/todo-app-ios/todo-app-ios/LazyView.swift).

**Reply by plusmobileapps** [reply]
Thanks for the response, `NavigationLink` does seem to go against the idea behind Decompose since it using SwiftUI's navigation. I did see that [iOS example on MVIKotlin](https://github.com/arkivanov/MVIKotlin/tree/master/sample/todo-app-ios/todo-app-ios) utilizes `NavigationLink` to achieve the native back swipe gesture, but I do like the concept behind Decompose consolidating routing logic. 

I wonder if there is another way to hook into the back swipe gesture on SwiftUI with Decompose without using `NavigationLink`? 🤔

**Reply by arkivanov** [reply]
Again, I have little experience with SwiftUI, but I would try the following. A while ago, when the TodoApp example was in this repository, it contained the [RouterView](https://github.com/arkivanov/Decompose/commit/8f26a4bd8e19ae93bff98a8f17cc5f60e7ba5a7b#diff-ef2df763bc6d06e28d3b943ff3dbffecd8284ae4db90d234c300ccacf740e81b). This view just renders all child views in `ZStack`, one on top another. This solved the issue with UI state preservation when navigating back (e.g. scroll position). So I would try to use this view for the swiping as well. The gesture recognition and swiping perhaps should be done using SwiftUI, hope there is a way of doing it. And when the top view is swiped out, just call a function on your component so it will pop the stack. Hope this helps.
